### PR TITLE
[NU-1324] Fragment and Built-in components definition preparation moved closer to the domain layer

### PR DESCRIPTION
**Pull request by arkadius**
## Describe your changes

Temporary degradations that will be fixed in the next PR:
- Icons for components with the same name but other type (e.g. `kafka`) will be wrong (always source or sink icon will be picked
- During component deduplication, we don't validate that deduplicated components have the same properties like icon etc. 

Non-temporary degradation:
- The mentioned deduplication above won't be done during model reload but during request for components - the previous approach didn't work correctly with the `AdditionalUIConfigProvider`

## Checklist before merge
- [ ] Related issue ID is placed at the beginning of PR title in \[brackets\] (can be GH issue or Nu Jira issue)
- [ ] Code is cleaned from temporary changes and commented out lines
- [ ] Parts of the code that are not easy to understand are documented in the code
- [ ] Changes are covered by automated tests
- [ ] Showcase in dev-application.conf added to demonstrate the feature
- [ ] Documentation added or updated
- [ ] Added entry in _Changelog.md_ describing the change from the perspective of a public distribution user
- [ ] Added _MigrationGuide.md_ entry in the appropriate subcategory if introducing a breaking change
- [ ] Verify that PR will be squashed during merge

### CODE REVIEW

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/component/NodeTemplateParameterPreparer.scala [hunk 1]
```diff
@@ -6,13 +6,13 @@ import pl.touk.nussknacker.engine.graph.expression.Expression
 
 object NodeTemplateParameterPreparer {
```
Comment (lciolecki) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/component/NodeTemplateParameterPreparer.scala:7]: NodeTemplateParameterPreparer -> NodeParameterTemplatsPreparer?

File: interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/component/defaultconfig/DefaultComponentConfigDeterminer.scala [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+      componentGroup = Some(configData.componentGroup),
+      componentId = None
+    )
+  }
+
+  def forBuiltInComponent(info: ComponentInfo): SingleComponentConfig = SingleComponentConfig(
+    params = None,
+    Some(DefaultsComponentIcon.forBuiltInComponent(info)),
+    // TODO: move from defaultModelConfig.conf to here + convention instead of code
+    docsUrl = None,
```
Comment (lciolecki) [inline interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/component/defaultconfig/DefaultComponentConfigDeterminer.scala:47]: WDYT e.g. https://nussknacker.io/documentation/docs/next/scenarios_authoring/BasicNodes/#filter? We have some documentation for BultIn / Basic components.

**Review by lciolecki — APPROVED** [review]
LGTM, change in nice direction.

File: docs/installation_configuration_guide/model/ModelConfiguration.md [hunk 3]
```diff
@@ -124,6 +124,15 @@ Example (see [dev application config](https://github.com/TouK/nussknacker/blob/s
   }
 ```
 
+As a key in the configuration you can use component name or full component identifier which is in the `componentType-componentName` format.
+The second approach is mostly useful when there is more than one component with the same name (e.g. `source-kafka` and `sink-kafka` components). 
+Available component types:
+- `source` - for components located in the `sources` toolbox panel
+- `sink` - for components located in the `sinks` toolbox panel
+- `service` - for components located in the `enrichers` and `services` toolbox panel
```
Comment (wrzontek) [inline docs/installation_configuration_guide/model/ModelConfiguration.md:132]:
```suggestion
- `service` - for components located in the `enrichers` and `services` toolbox panels
```
same for `custom`

**Review by wrzontek — COMMENTED** [review]
Very nice, the de-spaghettifying of UIProcessObjectsFactory is beatiful
mostly minor visual nitpicks in the review

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/definition/AdditionalUIConfigFinalizer.scala [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  ScenarioPropertyConfig,
+  SingleComponentConfig
+}
+import pl.touk.nussknacker.engine.api.process.ProcessingType
+import pl.touk.nussknacker.engine.definition.component.ComponentStaticDefinition
+import pl.touk.nussknacker.engine.definition.model.ModelDefinition
+import pl.touk.nussknacker.engine.util.Implicits.RichScalaMap
+import pl.touk.nussknacker.ui.component.{ComponentAdditionalConfigConverter, DefaultComponentIdProvider}
+
+class AdditionalUIConfigFinalizer(
```
Comment (wrzontek) [inline designer/server/src/main/scala/pl/touk/nussknacker/ui/definition/AdditionalUIConfigFinalizer.scala:16]: this class handles also base components, meaning that the first TODO in `AdditionalUIConfigProvider` can be removed 
` * TODO: The current implementation allows providing configs only for standard components - meaning that base components aren't handled.`

File: components-api/src/main/scala/pl/touk/nussknacker/engine/api/component/ComponentInfo.scala [hunk 5]
```diff
@@ -27,9 +27,12 @@ object BuiltInComponentInfo {
...
 
+  val FragmentDefinitionComponents: List[ComponentInfo] = List(FragmentInputDefinition, FragmentOutputDefinition)
+
   val AllAvailableForScenario: List[ComponentInfo] = List(Filter, Split, Choice, Variable, RecordVariable)
 
-  val All: List[ComponentInfo] =
-    AllAvailableForScenario ::: List(FragmentInputDefinition, FragmentOutputDefinition)
+  val All: List[ComponentInfo] = AllAvailableForScenario ::: FragmentDefinitionComponents
+
+  val AllAvailableForFragment: List[ComponentInfo] = All
```
[the remark this answers is not in the document]
Comment (arkadius) [reply components-api/src/main/scala/pl/touk/nussknacker/engine/api/component/ComponentInfo.scala:36]: In some places it is just easier to use `All` - I don't think that after a few weeks we would remember that both lists are equals.

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/component/ComponentIdProvider.scala [hunk 6]
```diff
@@ -39,21 +41,13 @@ class DefaultComponentIdProvider(configs: Map[String, ComponentsUiConfig]) exten
       .map(createComponentId(processingType, _))
 
   private def getOverriddenComponentId(
-      processingType: String,
-      componentName: String,
+      processingType: ProcessingType,
+      info: ComponentInfo,
```
[the remark this answers is not in the document]
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/component/ComponentIdProvider.scala:45]: I'll change `ComponentInfo` -> `ComponentId` in all places in the next PR

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/definition/scenarioproperty/FragmentPropertiesConfig.scala [hunk 7]
```diff
@@ -0,0 +1,21 @@
+package pl.touk.nussknacker.ui.definition.scenarioproperty
+
+import pl.touk.nussknacker.engine.api.FragmentSpecificData
+import pl.touk.nussknacker.engine.api.component.ScenarioPropertyConfig
+import pl.touk.nussknacker.engine.api.definition.StringParameterEditor
+
+object FragmentPropertiesConfig {
+
+  val docsUrlConfig: (String, ScenarioPropertyConfig) = FragmentSpecificData.docsUrlName ->
```
[the remark this answers is not in the document]
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/definition/scenarioproperty/FragmentPropertiesConfig.scala:9]: In this area of the code, we have everywhere it lowerCase, take a look at `FlinkStreamingPropertiesConfig.*`, `StreamMetaData.*`, `FragmentSpecificData.*`. I want to be consistent with these places. If you want to make it everwhere UpperCase, please submit a PR.

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/component/NodeTemplateParameterPreparer.scala (hunk 1 above)
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/component/NodeTemplateParameterPreparer.scala:7]: Good point

File: interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/component/defaultconfig/DefaultComponentConfigDeterminer.scala (hunk 2 above)
Comment (arkadius) [reply interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/component/defaultconfig/DefaultComponentConfigDeterminer.scala:47]: Take a look at TODO above, we already have these links in the `defaultModelConfig.conf`. I wanted to keep this change small. I'll fix it in the next PR

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/component/ComponentService.scala [hunk 8]
```diff
@@ -162,10 +163,10 @@ class DefaultComponentService(
         // We assume that fragments have unique component ids ($processing-type-fragment-$name) thus we do not need to validate them.
         val componentObjects = componentObjectsService.prepare(
           processingType,
-          processingTypeData,
+          processingTypeData._1,
+          processingTypeData._2,
```
[the remark this answers is not in the document]
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/component/ComponentService.scala:167]: I'd prefer it as well. It is temporary, I have plan to de-spaghettify `ComponentService` in the next change

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/component/ComponentsValidator.scala [hunk 9]
```diff
@@ -13,6 +13,7 @@ import pl.touk.nussknacker.ui.component.WrongConfigurationAttribute.{
   WrongConfigurationAttribute
 }
 
+// FIXME: rewrite to domain based approach, use it the ComponentService then
 private[component] object ComponentsValidator {
 
   def checkUnsafe(
```
[the remark this answers is not in the document]
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/component/ComponentsValidator.scala:19]: Yes, it's true. I'll fix it in the next PR

File: designer/server/src/test/scala/pl/touk/nussknacker/ui/component/DefaultComponentServiceSpec.scala [hunk 10]
```diff
@@ -525,49 +533,59 @@ class DefaultComponentServiceSpec
 
     forAll(testingData) {
       (user: LoggedUser, expectedComponents: List[ComponentListElement], possibleCategories: List[String]) =>
-        val components = defaultComponentService.getComponentsList(user).futureValue
+        val returnedComponents = defaultComponentService.getComponentsList(user).futureValue
 
-        def counts(list: List[ComponentListElement]) = list.map(el => el.id -> el.usageCount).toMap
+        returnedComponents.map(_.id).sortBy(_.value) should contain theSameElementsAs returnedComponents
```
[the remark this answers is not in the document]
Comment (arkadius) [reply designer/server/src/test/scala/pl/touk/nussknacker/ui/component/DefaultComponentServiceSpec.scala:538]: Good catch! Fixed

File: engine/flink/executor/src/main/scala/pl/touk/nussknacker/engine/process/compiler/StubbedFragmentSourceDefinitionPreparer.scala [hunk 11]
```diff
@@ -29,22 +29,20 @@ class StubbedFragmentSourceDefinitionPreparer(
...
-    parameters = Nil,
-    returnType = Some(fragmentReturnType),
-    categories = None,
-    componentConfig = SingleComponentConfig.zero,
-    componentTypeSpecificData = FragmentSpecificData
-  )
-
   def createSourceDefinition(frag: FragmentInputDefinition): MethodBasedComponentDefinitionWithImplementation = {
     val inputParameters = fragmentDefinitionExtractor.extractParametersDefinition(frag).value
+    val staticDefinition = FragmentComponentDefinition.prepareStaticDefinition(
```
[the remark this answers is not in the document]
Comment (arkadius) [reply engine/flink/executor/src/main/scala/pl/touk/nussknacker/engine/process/compiler/StubbedFragmentSourceDefinitionPreparer.scala:34]: We have a code duplication in two places. I needed to add `DefaultComponentConfigDeterminer.forFragment` invocation in both of them so I decided to extract it.

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/definition/AdditionalUIConfigFinalizer.scala [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+  def finalizeModelDefinition(
+      modelDefinition: ModelDefinition[ComponentStaticDefinition],
+      processingType: ProcessingType
+  ): ModelDefinition[ComponentStaticDefinition] = {
+    val componentIdProvider = new DefaultComponentIdProvider({ (_, info) =>
+      modelDefinition.getComponent(info).map(_.componentConfig)
+    })
+
+    val modelDefinitionWithIds = ModelDefinitionWithComponentIds(
```
[the remark this answers is not in the document]
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/definition/AdditionalUIConfigFinalizer.scala:28]: We need to discuss first, if we want to move this additional UI config enrichments into `ModelDefinitionExtractor`.  If so, I feel in my bones that this class will disappear without this refactor :)

**Pull request merged**